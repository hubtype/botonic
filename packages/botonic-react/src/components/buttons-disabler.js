import { Button } from '../components/button'
import { WEBCHAT } from '../constants'
import { isCarousel } from '../message-utils'
import { strToBool } from '../util/objects'
import { deepMapWithIndex } from '../util/react'
import { _getThemeProperty } from '../util/webchat'

export class ButtonsDisabler {
  static constructBrowserProps(props) {
    const disabledProps = { disabled: props.disabled }
    if (props.autodisable !== undefined)
      disabledProps.autodisable = strToBool(props.autodisable)
    if (props.disabledstyle !== undefined)
      disabledProps.disabledstyle = props.disabledstyle
    return disabledProps
  }
  static constructNodeProps(props) {
    const disabledProps = {}
    if (props.autodisable !== undefined)
      disabledProps.autodisable = String(props.autodisable)
    if (props.disabledstyle !== undefined)
      disabledProps.disabledstyle = JSON.stringify(props.disabledstyle)
    return disabledProps
  }

  static withDisabledProps(props) {
    return {
      disabled: props.disabled,
      autodisable: props.autodisable,
      disabledstyle: props.disabledstyle,
    }
  }

  static resolveDisabling(theme, props) {
    const getThemeProperty = _getThemeProperty(theme)
    const autoDisable =
      props.autodisable !== undefined
        ? props.autodisable
        : getThemeProperty(
            WEBCHAT.CUSTOM_PROPERTIES.buttonAutoDisable,
            WEBCHAT.DEFAULTS.BUTTON_AUTO_DISABLE
          )
    const computedDisabledStyle =
      props.disabledstyle !== undefined
        ? props.disabledstyle
        : getThemeProperty(WEBCHAT.CUSTOM_PROPERTIES.buttonDisabledStyle, {})

    const disabledStyle = {
      ...WEBCHAT.DEFAULTS.BUTTON_DISABLED_STYLE,
      ...computedDisabledStyle,
    }
    return { autoDisable, disabledStyle }
  }

  static updateChildrenButtons(children, additionalProps = undefined) {
    return deepMapWithIndex(children, n => {
      if (n.type === Button) return this.updateButtons(n, additionalProps)
      return n
    })
  }

  static updateButtons(node, additionalProps) {
    if (!additionalProps) additionalProps = {}
    else {
      additionalProps = {
        disabled:
          node.props.disabled === true
            ? node.props.disabled
            : additionalProps.disabled,
        setDisabled: additionalProps.setDisabled,
        parentId: additionalProps.parentId,
      }
    }
    return {
      ...node,
      props: {
        ...node.props,
        ...additionalProps,
      },
    }
  }

  static getUpdatedMessage(messageToUpdate, { autoDisable, disabledStyle }) {
    const updateMsgButton = button => {
      return {
        ...button,
        ...{
          disabled: true,
        },
      }
    }
    if (
      isCarousel(messageToUpdate) &&
      messageToUpdate.data &&
      messageToUpdate.data.elements
    ) {
      messageToUpdate.data.elements = messageToUpdate.data.elements.map(e => ({
        ...e,
        ...{
          buttons: e.buttons.map(updateMsgButton),
        },
      }))
      return messageToUpdate
    } else {
      return {
        ...messageToUpdate,
        ...{
          buttons: messageToUpdate.buttons.map(updateMsgButton),
        },
      }
    }
  }
}
