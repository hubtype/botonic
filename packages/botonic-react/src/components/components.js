/**
 * Get the name of an element (instantiated component)
 * @param {React.Element} component
 * @returns {string}
 */
export function getElementName(component) {
  return component.type.name
}

/**
 * Get the name of a component TYPE
 * @param {React.ComponentType} component
 * @returns {string}
 */
export function getComponentTypeName(component) {
  return component.displayName || component.name || 'Component'
}
