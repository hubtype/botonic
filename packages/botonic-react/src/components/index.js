export { Audio } from './audio'
export { Button } from './button'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Element } from './element'
export { Image } from './image'
export { Location } from './location'
export { MessageTemplate } from './message-template'
export * from './multichannel'
export { Pic } from './pic'
export { Reply } from './reply'
export { ShareButton } from './share-button'
export { Subtitle } from './subtitle'
export { Text } from './text'
export { Title } from './title'
export { Video } from './video'
export { WebchatSettings } from './webchat-settings'
export { WhatsappTemplate } from './whatsapp-template'
