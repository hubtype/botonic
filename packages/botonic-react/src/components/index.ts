export { Audio } from './audio'
export { Button } from './button'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Element } from './element'
export { Image } from './image'
export * from './index-types'
export { Location } from './location'
export { Message } from './message'
export { MessageTemplate } from './message-template'
export * from './multichannel'
export { Pic } from './pic'
export { Raw } from './raw'
export { Reply } from './reply'
export { ShareButton } from './share-button'
export { Subtitle } from './subtitle'
export { Text } from './text'
export { Title } from './title'
export { Video } from './video'
export { WebchatSettings } from './webchat-settings'
export {
  WhatsappButtonList,
  WhatsappButtonListProps,
  WhatsappButtonListRowProps,
  WhatsappButtonListSectionProps,
} from './whatsapp-button-list'
export {
  WhatsappCTAUrlButton,
  WhatsappCTAUrlButtonProps,
} from './whatsapp-cta-url-button'
export { WhatsappTemplate } from './whatsapp-template'
