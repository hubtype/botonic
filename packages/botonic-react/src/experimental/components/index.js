export { Button } from '../../components/button'
export { Element } from '../../components/element'
export { MessageTemplate } from '../../components/message-template'
export * from '../../components/multichannel'
export { Pic } from '../../components/pic'
export { Reply } from '../../components/reply'
export { ShareButton } from '../../components/share-button'
export { Subtitle } from '../../components/subtitle'
export { Title } from '../../components/title'
export { WebchatSettings } from '../../components/webchat-settings'

// Experimental
export { Audio } from './audio'
export { Carousel } from './carousel'
export { customMessage } from './custom-message'
export { Document } from './document'
export { Image } from './image'
export { Location } from './location'
export { Text } from './text'
export { Video } from './video'
export { WhatsappTemplate } from './whatsapp-template'
