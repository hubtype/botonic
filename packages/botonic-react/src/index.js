export { BotonicInputTester, BotonicOutputTester } from './botonic-tester'
export * from './components'
export { RequestContext, WebchatContext } from './contexts'
export { DevApp } from './dev-app'
export { msgsToBotonic, msgToBotonic } from './msg-to-botonic'
export { NodeApp } from './node-app'
export { staticAsset } from './util/environment'
export { getBotonicApp, Webchat } from './webchat/index.js'
export { WebchatApp } from './webchat-app'
export { WebviewApp } from './webview'
