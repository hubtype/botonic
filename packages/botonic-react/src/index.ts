const React = require('react')
const ReactDOMServer = require('react-dom/server')

import { default as i18n } from '@botonic/core/lib/i18n'

declare global {
    namespace JSX {
        interface IntrinsicElements {
            [elemName: string]: any;
        }
    }
}

export async function renderReactAction(req, action) {
	let props = {}
	if(action.botonicInit) {
       props = await action.botonicInit({req})
    }

    let component = React.createElement(action, (props && typeof(props)==='object') ? Object.assign(props, req) : req)
    return ReactDOMServer.renderToStaticMarkup(component)
}

export function setLocale(req: any, locale: string) {
    req.context.__locale = locale
    i18n.setLocale(locale)
}
export { BotonicWebview } from './webview'
export { MessageTemplate } from './messageTemplate'
export { ShareButton } from './shareButton'
