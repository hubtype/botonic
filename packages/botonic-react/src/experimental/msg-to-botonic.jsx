import React from 'react'

import {
  isAudio,
  isCarousel,
  isCustom,
  isDocument,
  isImage,
  isLocation,
  isText,
  isVideo,
} from '../../src/message-utils'
import { Button } from '../components/button'
import { ButtonsDisabler } from '../components/buttons-disabler'
import { Element } from '../components/element'
import { Pic } from '../components/pic'
import { Reply } from '../components/reply'
import { Subtitle } from '../components/subtitle'
import { Title } from '../components/title'
// Experimental
import { Audio } from './components/audio'
import { Carousel } from './components/carousel'
import { Document } from './components/document'
import { Image } from './components/image'
import { Location } from './components/location'
import { Text } from './components/text'
import { Video } from './components/video'

/**
 *
 * @param msg {object}
 * @param customMessageTypes {{customTypeName}[]?}
 * @return {React.ReactNode}
 */

export function msgToBotonic(msg, customMessageTypes) {
  console.log('calling msgToBotonic', { msg })
  delete msg.display
  if (isCustom(msg)) {
    try {
      return customMessageTypes
        .find(mt => mt.customTypeName === msg.customTypeName)
        .deserialize(msg)
    } catch (e) {
      console.log(e)
    }
  } else if (isText(msg)) {
    return (
      <Text {...msg} key={msg.key}>
        {msg.text}
        {msg.buttons && msg.buttons.length > 0 && buttonsParse(msg.buttons)}
        {msg.replies && msg.replies.length > 0 && quickrepliesParse(msg)}
      </Text>
    )
  } else if (isCarousel(msg)) {
    return (
      <Carousel {...msg} key={msg.key}>
        {elementsParse(msg.elements)}
      </Carousel>
    )
  } else if (isImage(msg)) {
    return (
      <Image key={msg.key} {...msg}>
        {msg.buttons && buttonsParse(msg.buttons)}
      </Image>
    )
  } else if (isVideo(msg)) {
    return <Video {...msg}>{msg.buttons && buttonsParse(msg.buttons)}</Video>
  } else if (isAudio(msg)) {
    return <Audio {...msg}>{msg.buttons && buttonsParse(msg.buttons)}</Audio>
  } else if (isDocument(msg)) {
    return (
      <Document {...msg}>{msg.buttons && buttonsParse(msg.buttons)}</Document>
    )
  } else if (isLocation(msg)) return <Location {...msg} />

  console.warn(`Not converting message of type ${msg.type}`)
  return null
}

function rndStr() {
  return Math.random().toString()
}

/**
 * @param msgs {object|object[]}
 * @param customMessageTypes {{customTypeName}[]?}
 * @return {React.ReactNode}
 */
export function msgsToBotonic(msgs, customMessageTypes) {
  if (Array.isArray(msgs)) {
    return (
      <>
        {msgs.map((msg, i) => {
          if (msg.key == null) {
            msg.key = `msg${i}`
          }
          return msgToBotonic(msg, customMessageTypes)
        })}
      </>
    )
  }
  return msgToBotonic(msgs, customMessageTypes)
}

function elementsParse(elements) {
  return elements.map((e, i) => (
    <Element key={i}>
      <Pic src={e.pic} />
      <Title>{e.title}</Title>
      <Subtitle>{e.subtitle}</Subtitle>
      {e.buttons && buttonsParse(e.buttons)}
    </Element>
  ))
}

function buttonsParse(buttons) {
  return buttons.map((b, i) => {
    let payload = b.payload
    if (b.path) payload = `__PATH_PAYLOAD__${b.path}`
    const url = b.messenger_extensions ? null : b.url
    const target = b.messenger_extensions ? null : b.target
    const title = b.title
    const webview = b.messenger_extensions ? b.url : b.webview
    const disabledProps = ButtonsDisabler.constructBrowserProps(b)
    return (
      <Button
        key={i}
        payload={payload}
        url={url}
        target={target}
        webview={webview}
        {...disabledProps}
      >
        {title}
      </Button>
    )
  })
}

function quickrepliesParse(msg) {
  let replies = null
  if (msg.replies) {
    replies = msg.replies.map((el, i) => {
      let payload = el.payload
      if (el.path) payload = `__PATH_PAYLOAD__${el.path}`
      return (
        <Reply key={i} payload={payload}>
          {el.title}
        </Reply>
      )
    })
  }
  return replies
}
