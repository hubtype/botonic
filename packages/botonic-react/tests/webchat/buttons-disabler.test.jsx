import React from 'react'
import TestRenderer from 'react-test-renderer'

import {
  Button,
  Carousel,
  Element,
  Pic,
  Subtitle,
  Text,
  Title,
} from '../../src/components'
import { ButtonsDisabler } from '../../src/components/buttons-disabler'
import { msgToBotonic } from '../../src/msg-to-botonic'

const renderToJSON = sut => TestRenderer.create(sut).toJSON()

describe('TEST: ButtonsDisabler (Disabling buttons in Webchat)', () => {
  it('Appends the disabling props to a component', () => {
    const componentProps = (
      <Button payload='some-payload' autodisable={false}>
        Button Text
      </Button>
    ).props
    const sut = {
      ...componentProps,
      ...ButtonsDisabler.withDisabledProps(componentProps),
    }
    expect(sut).toEqual({
      payload: 'some-payload',
      autodisable: false,
      children: 'Button Text',
      disabled: undefined,
      disabledstyle: undefined,
    })
  })
  it('Passes a message with expected props to node (text with buttons)', () => {
    const sut = COMPONENTS.TEXT_WITH_BUTTONS
    const tree = renderToJSON(sut)
    expect(tree).toMatchSnapshot()
  })
  it('Passes a message with expected props to node (carousel)', () => {
    const sut = COMPONENTS.CAROUSEL
    const tree = renderToJSON(sut)
    expect(tree).toMatchSnapshot()
  })
  it('Fills the props correctly in browser', () => {
    const props = { autodisable: false }
    const sut = ButtonsDisabler.constructBrowserProps(props)
    expect(sut).toEqual({ disabled: undefined, autodisable: false })
  })
  it('Fills the props correctly in node', () => {
    const props = { autodisable: false }
    const sut = ButtonsDisabler.constructNodeProps(props)
    expect(sut).toEqual({ autodisable: 'false' })
  })
  it('Resolves disabling properties correctly', () => {
    const props = (
      <Button payload='some-payload' autodisable={false}>
        Button Text
      </Button>
    ).props
    const sut = ButtonsDisabler.resolveDisabling(WEBCHAT.theme, props)
    expect(sut.autoDisable).toEqual(false)
    expect(sut.disabledStyle).toEqual({
      opacity: 0.5,
      cursor: 'auto',
      pointerEvents: 'none',
      backgroundColor: 'green',
    })
  })
  it('Updates children with buttons disabling properties', () => {
    const initialChildren = COMPONENTS.TEXT_WITH_BUTTONS.props.children
    const sut = ButtonsDisabler.updateChildrenButtons(initialChildren, {
      parentId: '1234',
      disabled: true,
      setDisabled: () => {},
    })
      .filter(e => e.type === Button)
      .map(e => e.props)
    expect(sut[0].autodisable).toEqual(false)
    expect(sut[0].disabledstyle).toEqual(undefined)
    expect(sut[0].disabled).toEqual(true)
    expect(sut[1].autodisable).toEqual(true)
    expect(sut[1].disabledstyle).toEqual({ backgroundColor: 'red' })
    expect(sut[1].disabled).toEqual(true)
  })
  it('Converts correctly a buttonmessage event', () => {
    const buttonMessage = {
      type: 'buttonmessage',
      text:
        'Here I display two types of buttons, the first one is a URL button and the second is a payload button:',
      buttons: [
        {
          type: 'postback',
          title: 'Visit botonic.io',
          payload: 'https://botonic.io',
          autodisable: 'false',
        },
        {
          type: 'postback',
          title: 'Show me a carousel',
          payload: 'carousel',
        },
      ],
      payload: {},
      id: '16972707e231427fbeeb707ab79b0b19',
    }
    const sut = msgToBotonic(buttonMessage)
    const buttons = sut.props.children.props.children[1]
    const button1Props = buttons[0].props
    const button2Props = buttons[1].props
    expect(button1Props.autodisable).toEqual(false)
    expect(button2Props.autodisable).toEqual(undefined)
  })
  it('Converts correctly a carousel event', () => {
    const carouselMessage = {
      type: 'carousel',
      elements: [
        {
          title: 'ELEMENT1',
          buttons: [
            {
              type: 'postback',
              title: 'Visit website',
              payload: 'URL1',
            },
          ],
          image_url: 'PIC1',
          subtitle: 'DESC1',
        },
        {
          title: 'ELEMENT2',
          buttons: [
            {
              type: 'postback',
              title: 'KAKA',
              payload: 'URL2',
              autodisable: 'true',
              disabledstyle: {
                backgroundColor: 'red',
              },
            },
          ],
          image_url: 'PIC2',
          subtitle: 'DESC2',
        },
      ],
      payload: {},
      id: 'cca3f3a852f6412885ecd9149a89deee',
    }
    const sut = msgToBotonic(carouselMessage)
    const elements = sut.props.children
    const element1ButtonProps = elements[0].props.children[3][0].props
    const element2ButtonProps = elements[1].props.children[3][0].props

    expect(Object.keys(element1ButtonProps)).toEqual([
      'payload',
      'url',
      'target',
      'webview',
      'disabled',
      'children',
    ])
    expect(Object.keys(element2ButtonProps)).toEqual([
      'payload',
      'url',
      'target',
      'webview',
      'disabled',
      'autodisable',
      'disabledstyle',
      'children',
    ])
    expect(element2ButtonProps.disabled).toEqual(undefined)
    expect(element2ButtonProps.autodisable).toEqual(true)
    expect(element2ButtonProps.disabledstyle).toEqual({
      backgroundColor: 'red',
    })
  })
})

const WEBCHAT = {
  theme: {
    style: {
      width: 500,
    },
    button: {
      disabledstyle: {
        opacity: 0.5,
        cursor: 'auto',
        pointerEvents: 'none',
        backgroundColor: 'green',
      },
      autodisable: true,
    },
  },
}

const COMPONENTS = {
  TEXT_WITH_BUTTONS: (
    <Text>
      Here I display two types of buttons, the first one is a URL button and the
      second is a payload button:
      <Button payload='https://botonic.io' autodisable={false}>
        Visit botonic.io
      </Button>
      <Button
        payload='carousel'
        autodisable={true}
        disabledstyle={{ backgroundColor: 'red' }}
      >
        Show me a carousel
      </Button>
    </Text>
  ),
  CAROUSEL: (
    <Carousel>
      {[
        {
          name: 'ELEMENT1',
          desc: 'DESC1',
          url: 'URL1',
          pic: 'PIC1',
          props: {
            payload: 'payload1',
            children: 'text1',
          },
        },
        {
          name: 'ELEMENT2',
          desc: 'DESC2',
          url: 'URL2',
          pic: 'PIC2',
          props: {
            payload: 'payload2',
            autodisable: true,
            disabledstyle: { backgroundColor: 'red' },
            children: 'text2',
          },
        },
      ].map((e, i) => (
        <Element key={e.name}>
          <Pic src={e.pic} />
          <Title>{e.name}</Title>
          <Subtitle>{e.desc}</Subtitle>
          <Button {...e.props} />
        </Element>
      ))}
    </Carousel>
  ),
}
