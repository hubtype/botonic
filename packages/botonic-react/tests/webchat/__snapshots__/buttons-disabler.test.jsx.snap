// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST: ButtonsDisabler (Disabling buttons in Webchat) Passes a message with expected props to node (carousel) 1`] = `
<message
  delay={0}
  style={
    Object {
      "backgroundColor": "rgba(0, 0, 0, 0)",
      "padding": 0,
      "width": "85%",
    }
  }
  type="carousel"
  typing={0}
>
  <element>
    <pic>
      PIC1
    </pic>
    <title>
      ELEMENT1
    </title>
    <desc>
      DESC1
    </desc>
    <button
      payload="payload1"
    >
      text1
    </button>
  </element>
  <element>
    <pic>
      PIC2
    </pic>
    <title>
      ELEMENT2
    </title>
    <desc>
      DESC2
    </desc>
    <button
      autodisable="true"
      disabledstyle="{\\"backgroundColor\\":\\"red\\"}"
      payload="payload2"
    >
      text2
    </button>
  </element>
</message>
`;

exports[`TEST: ButtonsDisabler (Disabling buttons in Webchat) Passes a message with expected props to node (text with buttons) 1`] = `
<message
  delay={0}
  markdown={1}
  type="text"
  typing={0}
>
  Here I display two types of buttons, the first one is a URL button and the second is a payload button:
  <button
    autodisable="false"
    payload="https://botonic.io"
  >
    Visit botonic.io
  </button>
  <button
    autodisable="true"
    disabledstyle="{\\"backgroundColor\\":\\"red\\"}"
    payload="carousel"
  >
    Show me a carousel
  </button>
</message>
`;
